06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+```
+
+Finally, in order to search over the subsets of predictors of *all* sizes,
+we will need to iterate over the possible sizes from 1 to 4.
+If you recall the end of the wrangling chapter, we mentioned
+that sometimes one needs more flexible forms of iteration than what 
+we have used earlier, and in these cases one typically resorts to
+[a for loop](https://r4ds.had.co.nz/iteration.html#iteration).
+This is one of those cases! Here we will use two for loops:
+one over the possible sizes of the predictor subset
+(where you see `for (i in 1:length(names))` below)
+and another over the subsets of each size (where you see `for (j in 1:length(combs))` below).
+For each subset, we construct a model formula,
+pass it into a `recipe`, build a `workflow` that tunes
+a K-NN classifier using 5-fold cross-validation, 
+and finally records the estimated accuracy.
+
+```{r 06-best-subset-3, warning = FALSE}
+set.seed(1)
+
+# create an empty tibble to store the results
+accuracies <- tibble(size = integer(), index = integer(), model_string = character(), accuracy = numeric())
+
+# create a model specification
+knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
+     set_engine("kknn") %>%
+     set_mode("classification")
+
+# create a 5-fold cross-validation object
+cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
+
+# for every size from 1 to length(names)
+for (i in 1:length(names)) {
+    # get all combinations of predictors of size i
+    combs <- combn(names, i, simplify=FALSE)
+    
+    # for every combination of predictors of size i
+    for (j in 1:length(combs)) {
+        # create a model string for this combination of predictors
+        model_string <- paste("Class", "~", paste(combs[[j]], collapse="+"))
+
+        # create a recipe from the model string
+        cancer_recipe <- recipe(as.formula(model_string), data = cancer_subset) %>%
+            step_scale(all_predictors()) %>%
+            step_center(all_predictors())
+
+  
+        # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
+        acc <- workflow() %>%
+          add_recipe(cancer_recipe) %>%
+          add_model(knn_spec) %>%
+          tune_grid(resamples = cancer_vfold, grid = 10) %>%
+          collect_metrics() %>%
+          filter(.metric == "accuracy") %>%
+          summarize(mx = max(mean))
+        acc <- acc$mx %>% unlist()
+
+        # add this result to the dataframe
+        accuracies <- accuracies %>% add_row(size = i, index = j, model_string = model_string, accuracy = acc)
+    }
+}
+
+accuracies
+```
+The maximum accuracy estimate is obtained for the formula
+`Class ~ Concavity + Symmetry + Area`, so the best subset selection 
+method would result in a model with these three predictors. Note,
+however, that the model with all 4 predictors, as well as the 
+model `Class ~ Smoothness + Concavity + Area` are reasonable
+alternatives that provide a similar estimated accuracy.
```

Comment:
This result - that we should include all predictors doesn't bring home the message we started this chapter out with... I think we should see if there is an irrelevant predictor in that data set (although, from what I know of it, I don't think there is) or add in 1-2 of the irrelevant predictors and show that best subset selection will in fact remove them. Right now as the chapter is, we haven't demonstrated that, just told them that it will...

Reply:
fixed now!